docs/design/branch-mode.md

```diff
@@ -0,0 +1,117 @@
+# Branch mode
+
+This is a proposed configuration option to replace
+`experimental-advance-branches`. It is not intended to replace or obviate
+topics.
+
+The design below assumes that the configuration option is enabled. There should
+be no change from current behavior if the option is not enabled.
+
+**"Non-headless"** or **"branch"** mode: a git-colocated repository with the
+new configuration option enabled is considered to be in this mode whenever
+`.git/HEAD` (hereafter `HEAD`) is populated with a valid `jj` bookmark, and the
+bookmark is currently at `@` or `@-`. In this design, the bookmark will be
+referred to as `b(HEAD)`.
+
+**Note:** The only part of this design that is specific to `git` colocation is
+the use of `.git/HEAD`. The design could be implemented for any backend that
+has a concept of a "current" branch.
+
+## Invariants
+
+The configuration option will never affect any of these `jj` behaviors:
+
+* how `@` is updated by each `jj` command
+* the way commit topology or contents (ignoring bookmarks) evolve with each
+  command
+
+## Diagrams
+
+These Excalidraw diagrams show the effect of various `jj` commands when in
+branch mode:
+https://excalidraw.com/#json=pkmTWTuzkToGdc4xSUy2C,7uhU9FBI6rJK-U_XGCH4HA
+
+In these diagrams, a colored arrow is a `jj` bookmark, and a bold and colored
+commit-graph node is `b(HEAD)`.
+
+## Entering branch mode
+
+These commands always enter branch mode, by setting `.git/HEAD` appropriately:
+
+```
+git init
+git clone
+bookmark set <name> @
+bookmark create <name> @
+bookmark set <name> @-
+bookmark create <name> @-
+```
+
+See also "Changing branches."
+
+## Staying on the same Git branch
+
+When in branch mode, these operations preserve branch mode, and do not change
+which bookmark `HEAD` points to:
+
+```
+rebase
+abandon
+commit
+new [without specifying revisions]
+duplicate
+backout
+```
```

Comment:
None of these commands currently do anything special w.r.t. branches - it's all generic behavior that's implemented at a lower level. I'd really like to keep it that way. That ensures that the behavior is consistent between commands, including for our custom commands at Google, and in operations implemented by various UIs that use the library.

So, can you instead describe the behavior in a way that doesn't refer to a specific command (other than as an example)?

Reply:
I'll try to rephrase, but the short version is that these commits _aren't_ doing anything "special." If you've got an active branch, then that branch stays active, and the corresponding jj bookmark moves appropriately, i.e. it moves together with `@`.